projects/ADuCM3029_demo_cn0428_cn0429/src/cli.c

```diff
@@ -0,0 +1,986 @@
+*
+*Copyright 2015-2018(c) Analog Devices, Inc.
+*
+*All rights reserved.
+*
+*Redistribution and use in source and binary forms, with or without modification,
+*are permitted provided that the following conditions are met:
+*    - Redistributions of source code must retain the above copyright
+*      notice, this list of conditions and the following disclaimer.
+*    - Redistributions in binary form must reproduce the above copyright
+*      notice, this list of conditions and the following disclaimer in
+*      the documentation and/or other materials provided with the
+*      distribution.
+*    - Neither the name of Analog Devices, Inc. nor the names of its
+*      contributors may be used to endorse or promote products derived
+*      from this software without specific prior written permission.
+*    - The use of this software may or may not infringe the patent rights
+*      of one or more patent holders.  This license does not release you
+*      from the requirement that you obtain separate licenses from these
+*      patent holders to use this software.
+*    - Use of the software either in source or binary form, must be run
+*      on or directly connected to an Analog Devices Inc. component.
+*
+*THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+*INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT, MERCHANTABILITY AND FITNESS FOR A
+*PARTICULAR PURPOSE ARE DISCLAIMED.
+*
+*IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+*EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, INTELLECTUAL PROPERTY
+*RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+*BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+*STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+*THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************
+**/
+
+/***************************** Include Files **********************************/
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <ctype.h>
+#include "cli.h"
+#include "ADuCM3029_demo_cn0428_cn0429.h"
+#include "Ringbuffer.h"
+
+/********************************* Definitions ********************************/
+
+
+
+/************************** Variable Definitions ******************************/
+/* Available commands */
+char *CmdCommands[] = {
+	"",
+	"help",
+	"defaultsensor",
+	"sensorsconnected",
+	"readconfigs",
+
+	"readtemp",
+	"readhum",
+	"setmeastime",
+	"startmeas",
+	"stopmeas",
+	"setrtia",
+	"setrload",
+	"setvbias",
+	"setsensitivity",
+	"runeis",
+	"readeis",
+	"readeisfull",
+	"readrcal",
+	"runpulse",
+	"readpulse",
+	"pulseamplitude",
+	"pulseduration",
+	"readsensors",
+	"stopread",
+	"setupdaterate",
+	""
+};
+
+/* Functions for available commands */
+cmdFunc CmdFun[] = {
```

Comment:
this type and the `char *CmdCommands[] ` could be constructed like a table of strings to function pointers, where you can lookup the function pointer by string matching

something like  [this code is an example; it needs some adaptation ]
```
struct func_map {
     const char * name;
     cmdFunc  func; 
}
```

this has 2 benefits:
1. you don't need to export each function in the `cli.h` header
2. you can avoid multiple string matchings, like the one in `cli.c` + the one in `ADuCM3029_demo_cn0428_cn0429.cpp` (i.e. the check_string() function) ; though this would need a bit of re-design;

Reply:
I like this idea in general. But this is not what was done in the CLI example project or cn0326. Multiple string matchings could have been avoided by processing in the CLI file, but I would like to avoid unnecessary structural redesigns.